### ISSUE TITLE
Explore adding support for a file history (or file timeline) view

### ISSUE THREAD

**Issue by eamodio** [microsoft/vscode#84297]
# Goals

Add support for a unified file-based timeline view to be added to the Explorer sidebar which will track the active document (similar to the Outline view). This new view can be contributed by multiple sources, e.g. save/undo points, source control commits, test runs/failures, etc. Events from all sources will be aggregated into a single chronologically-ordered view. 

# Proposal

```ts
	export class TimelineItem {
		/**
		 * A date for when the timeline item occurred
		 */
		date: Date;

		/**
		 * A human-readable string describing the source of the timeline item. This can be used for filtering by sources so keep it consistent across timeline item types.
		 */
		source: string;

		/**
		 * Optional method to get the children of the timeline item (if any).
		 *
		 * @return Children of the timeline item (if any).
		 */
		getChildren?(): ProviderResult<TreeItem[]>;

		/**
		 * A human-readable string describing the timeline item. When `falsy`, it is derived from [resourceUri](#TreeItem.resourceUri).
		 */
		label: string;

		/**
		 * Optional id for the timeline item. See [TreeItem.id](#TreeItem.id) for more details.
		 */
		id?: string;

		/**
		 * The icon path or [ThemeIcon](#ThemeIcon) for the timeline item. See [TreeItem.iconPath](#TreeItem.iconPath) for more details.
		 */
		iconPath?: string | Uri | { light: string | Uri; dark: string | Uri } | ThemeIcon;

		/**
		 * A human readable string describing less prominent details of the timeline item. See [TreeItem.description](#TreeItem.description) for more details.
		 */
		description?: string | boolean;

		/**
		 * The [uri](#Uri) of the resource representing the timeline item (if any). See [TreeItem.resourceUri](#TreeItem.resourceUri) for more details.
		 */
		resourceUri?: Uri;

		/**
		 * The tooltip text when you hover over the timeline item.
		 */
		tooltip?: string | undefined;

		/**
		 * The [command](#Command) that should be executed when the timeline item is selected.
		 */
		command?: Command;

		/**
		 * [TreeItemCollapsibleState](#TreeItemCollapsibleState) of the timeline item.
		 */
		collapsibleState?: TreeItemCollapsibleState;

		/**
		 * Context value of the timeline item.  See [TreeItem.contextValue](#TreeItem.contextValue) for more details.
		 */
		contextValue?: string;

		/**
		 * @param label A human-readable string describing the timeline item
		 * @param date A date for when the timeline item occurred
		 * @param source A human-readable string describing the source of the timeline item
		 * @param collapsibleState [TreeItemCollapsibleState](#TreeItemCollapsibleState) of the timeline item. Default is [TreeItemCollapsibleState.None](#TreeItemCollapsibleState.None)
		 */
		constructor(label: string, date: Date, source: string, collapsibleState?: TreeItemCollapsibleState);
	}

	export interface TimelimeAddEvent {

		/**
		 * An array of timeline items which have been added.
		 */
		readonly items: readonly TimelineItem[];

		/**
		 * The uri of the file to which the timeline items belong.
		 */
		readonly uri: Uri;
	}

	export interface TimelimeChangeEvent {

		/**
		 * The date after which the timeline has changed. If `undefined` the entire timeline will be reset.
		 */
		readonly since?: Date;

		/**
		 * The uri of the file to which the timeline changed.
		 */
		readonly uri: Uri;
	}

	export interface TimelineProvider {
		onDidAdd?: Event<TimelimeAddEvent>;
		onDidChange?: Event<TimelimeChangeEvent>;

		/**
		 * Provide [timeline items](#TimelineItem) for a [Uri](#Uri) after a particular date.
		 *
		 * @param uri The uri of the file to provide the timeline for.
		 * @param since A date after which timeline items should be provided.
		 * @param token A cancellation token.
		 * @return An array of timeline items or a thenable that resolves to such. The lack of a result
		 * can be signaled by returning `undefined`, `null`, or an empty array.
		 */
		provideTimeline(uri: Uri, since: Date, token: CancellationToken): ProviderResult<TimelineItem[]>;
	}

	export namespace workspace {
		/**
		 * Register a timeline provider.
		 *
		 * Multiple providers can be registered. In that case, providers are asked in
		 * parallel and the results are merged. A failing provider (rejected promise or exception) will
		 * not cause a failure of the whole operation.
		 *
		 * @param selector A selector that defines the documents this provider is applicable to.
		 * @param provider A timeline provider.
		 * @return A [disposable](#Disposable) that unregisters this provider when being disposed.
		 */
		export function registerTimelineProvider(selector: DocumentSelector, provider: TimelineProvider): Disposable;
	}
```

A timeline source (e.g. an extension) registers a `TimelineProvider` for a set of documents. VS Code will then call the all the registered `TimelineProvider.provideTimeline` callbacks (in parallel) when the active editor changes and matches those registrations. The results will be merged into a unified set ordered by the `TimelineItem.date` and displayed in a new _File Timeline_ view in the Explorer sidebar.

A timeline provider can signal that new events have occurred via the `onDidAdd` event, providing the set of additional timeline items. A provider can also signal a refresh of its timeline items via the `onDidChange` event.

# Questions & Challenges

### API
- Should we provides a model/signal that a timeline provider should not be cached or more accurately flushed when the file is no longer the active editor? Basically an alternative to sending events, just call `provideTimeline` again
- Caching will be a bit challenging
  - How much do we keep around? And for how long?
- Should the `TimelimeChangeEvent` event provide a way to signal that an individual item(s) should be updated?
- Do we need a throttle on `onDidAdd`?
- Need to be careful with `id`s for the tree items, since they can come from multiple extensions. Probably should ensure a prefix or something per provider for any provided `id`s (or not give control over the `id`s at all) 

### Behavior
- Should tracking the active document be a toggle (like in GitLens where you can turn on/off tracking in the file history view on demand).
- Should there be a way to trigger a specific file/folder/uri timelime to be shown (which would also turn off active file tracking) -- again similar to GitLens
- Should the view support a refresh action that will drop timeline caches and re-request the timeline from all providers? (Hopefully we don't really need this)
- Should we support filtering based on the `source` of the timeline items?
/cc @jrieken 

Refs: https://github.com/microsoft/vscode/issues/83995

**Comment by eamodio** [comment]
# Latest Proposal

https://github.com/microsoft/vscode/blob/f2cc6fd3d8591b0754e58256ca3f97ca58c239a1/src/vs/vscode.proposed.d.ts#L1473-L1566

# Notes

- I am currently avoiding all the caching complexities outlined in the original post by making the [`TimelineService`](https://github.com/microsoft/vscode/blob/f2cc6fd3d8591b0754e58256ca3f97ca58c239a1/src/vs/workbench/contrib/timeline/common/timelineService.ts) not provide any caching at all. So as the active document changes, new requests are made to all registered timeline providers.

# Questions

- Currently, a `TimelineProvider` can signal that it is `replaceable`, but I think I should probably remove that ability and instead just deal with replacing existing providers by filtering out sources.
- Should we have the ability to show a timeline for a non-file Uri -- say a workspace, folder, or a git repo? -- Could be very powerful

# TODO

- Add support for the `DocumentScheme` when registering a `TimelineProvider` isn't currently hooked up -- need to move the code around to marshal `DocumentScheme`
- Add support for custom commands/menus to be applied to timeline items (using the `contextValue`)
- Add filtering support by `source` in the UI (and maybe backed by a setting)
- Add paging support -- investigate some sort of cursor support (something that the provider can control)
- Add ability to show a timeline for a specific file (say via a command on explorer files)
- Add ability to "pin/freeze" a timeline from following the active editor

**Comment by lukaszpolowczyk** [comment]
@eamodio Will this API give access to click history in the document, selection history in the document?

I use this in my extension, but I have to remember the clicked places myself.
And I know that VSCode has a history undo function (Go> Back) but also a selection history.

**Comment by eamodio** [comment]
@lukaszpolowczyk This API is more about allowing a source to provide timeline information for a given Uri (document for now, but probably folder and workspace in the future). And that provided information is then shown as a list in a new _Timeline_ view. This will enable extensions (and core) to start providing different sources to the timeline -- which could be undo history, selection history, etc. 

Although, in this first iteration we will be providing Git file history in the timeline.

**Comment by eamodio** [comment]
Also for this iteration this will only be available in insiders and you will need to add the following to your settings.json: `"timeline.showView": true`

**Comment by gjsjohnmurray** [comment]
@eamodio wrote in the TODO above:

> Add filtering support by `source` in the UI (and maybe backed by a setting)

Maybe worth the API comment for `source` recommending that an extension contributing to a timeline uses their extension ID (i.e. `<publisher>.<name>` from their manifest) as `source` or as a prefix for it, perhaps separated by /. For example: `acmecorp.dynamite` or for greater granularity `acmecorp.dynamite/major`, `acmecorp.dynamite/minor`.

Though if the timeline can already obtain the ID of the extension implementing the TimelineProvider that gave it the items it won't be necessary to ask providers to use `source`s that are structured so as to avoid conflicts.

**Comment by eamodio** [comment]
@gjsjohnmurray Yeah, I'm not sure if the `source` needs to be namespaced per-extension (though if we do end up wanting that, I would just do it internally). Currently with the `replaceable` property the desire is to be able to override a `source` with another one. But I still need to think more about that -- since I might just get rid of `replaceable` completely and rely solely on filtering (rather than overriding) and if so, then namespacing is probably appropriate.

**Comment by Zalastax** [comment]
@amelio-vazquez-reina asks if this could be used to support a tree based undo history https://github.com/microsoft/vscode/issues/20889#issuecomment-583194444:

> Could the the **new** **[Timeline view](https://code.visualstudio.com/updates/v1_42#_preview-features)** be used to support this feature? Or does it only support "[path graphs](https://en.wikipedia.org/wiki/Path_graph)" (aka "linear graphs"), i.e. graphs where nodes can have at most **one predecessor** and at most **one successor**?
> 
> Note that the "undo graph" (aka "history graph" aka "undo branches") is usually **at least** a tree, if not a general graph (if we assume a graph where any nodes with the same text collapse into one node)

**Comment by eamodio** [comment]
@Zalastax @amelio-vazquez-reina Currently the timeline is linear, but we are thinking about allowing each linear node to expand to contain a set of child nodes. Although the current thinking is that there will only be 1 level of children. 

For example, if you were to show the Git history of a folder, then each linear item would be a commit, which would expand to show the set of files within that commit.

**Comment by amelio-vazquez-reina** [comment]
Thanks @eamodio - What would you say is the main argument *against* supporting history trees in the Timeline view? is it the complexity and effort that it may require to do so **while** properly delivering a great / easy / intuitive UX for the user? Or are there any other reasons? 

In my mind, there are good UX designs that could accommodate tree views and make them optional.

In case it helps, below is a screenshot and animation of a similar feature (for navigating the full undo history of a file) in Emacs:

### Screenshot:
![4vbd9](https://user-images.githubusercontent.com/1993919/74043472-6b75e800-4997-11ea-8072-05e1a1e4dc11.png)

### Video example

![ezgif-2-458901b2fc48](https://user-images.githubusercontent.com/1993919/74043890-2a320800-4998-11ea-8804-eed4cd9b202f.gif)

**Comment by eamodio** [comment]
I can't say there is a specific argument against any usage, but we do want to keep things consistent and intuitive from a UX perspective, and avoid unnecessary complexity (in both the UI/UX as well as the API). We also want to make sure the Timeline view is accessible and not overwhelming especially when mixing different sources. At the same time, we are certainly open to evolving the UI/UX/API in response to new valuable use-cases.

For _history trees_, how would you envision that would look and behave? What would need to change to support it?

Also I'm sure I'm missing something, but I don't understand what you are trying to show in the video example.

**Comment by amelio-vazquez-reina** [comment]
Thanks @eamodio Very helpful.

> For history trees, how would you envision that would look and behave? What would need to change to support it?

Sorry the gif I originally uploaded was incomplete. I have now fixed it in my original comment. I'll describe what this gif shows, since that's probably the best to answer this question.

In the animation the user starts typing text on a buffer (editor in VScode), then undoing some of the typing, and then typing text again. At a specific point in time the user invokes a command called "undo-tree" (equivalent to our Timeline view) which splits then screen in half and shows, in the top half: the original file, and in the lower half: the current **history (undo) tree** of that file.

Once the undo tree is visible, the user switches focus to the tree and the user navigates up and down the history of the original file via keyboard shortcuts. When the focus is on the tree, the user can also **switch branches** in the tree when the user is sitting on a node with multiple children, via a shortcut (think left / right).

Most importantly, as the user is visiting different nodes, the original file (at the top) is updating its contents to reflect the state of the file corresponding to that node, and at any point in time, the user can switch focus to the original file in the top half, and continue editing from that state.

Does that make sense? I hope that helps!


One way is the following: 
* Every file technically speaking has a history (undo) tree. 
* Each node in the tree represents a state of the file. This can be a saved state, or an automatically snapshotted state (how Emacs and Vim do it). 
* As the user types and adds text, the user is just growing a path. 
* When the user does an `undo` operation, the user is effectively going back in the history tree (on a path, one node back for each undo operation). 
* If the user starts typing again after an undo, the user is then creating a new branch in the tree. 
* At any point in time the user can navigate the tree by e.g. clicking on a given node, or via keyboard shortcuts (undo-tree in Emacs allows the user to use C-n or C-p to go up/down in the tree, and bind a key to switching "active descending branch" for going down on a different branch)

That's what the video shows. The user is typing, then undoing

**Comment by zabel-xyz** [comment]
Good idea for such a feature ! I made my own tree in my extension [local-history](https://github.com/zabel-xyz/local-history). I'll contribute to this timeline view...

**Comment by JohnstonCode** [comment]
This is a nice feature. Not ran into any issues using it for svn commit history

**Comment by eamodio** [comment]
Here is the latest proposal:

https://github.com/microsoft/vscode/blob/604fb7637f58fb3d118fcf2460b800836494fd5d/src/vs/vscode.proposed.d.ts#L1814-L1966

`TimelineOptions` is now much simpler:

```ts
export interface TimelineOptions {
	/**
	 * A provider-defined cursor specifying the starting point of the timeline items that should be returned.
	 */
	cursor?: string;

	/**
	 * An optional maximum number timeline items or the all timeline items newer (inclusive) than the timestamp or id that should be returned.
	 * If `undefined` all timeline items should be returned.
	 */
	limit?: number | { timestamp: number; id?: string };
}
```

These are the usage combinations:
- `{ limit: number }` — to get the first page (most recent)
- `{ limit: { timestamp: number; id?: string }` — to get the most recent items, up to a set timeline item (or timestamp) -- we use this for providers with mutable histories (think git -- reset, rebase, etc), so when they tell us to update the timeline we can ask for all our current visible data
- `{ cursor: string, limit: number }` — to get the next page of data starting at a specific point (cursor)
- `{ }` — to get all items in one shot (no paging)
- `{ cursor: string }` — to get all items in one shot starting from a specific point (we don't have to use this, since we can just use the above to get everything again).

I've also simplified the `paging` property of the `Timeline`:

```ts
export interface Timeline {
	readonly paging?: {
		/**
		 * A provider-defined cursor specifying the starting point of timeline items which are after the ones returned.
		 * Use `undefined` to signal that there are no more items to be returned.
		 */
		readonly cursor: string | undefined;
	}

	/**
	 * An array of [timeline items](#TimelineItem).
	 */
	readonly items: readonly TimelineItem[];
}
```

Now you only return `paging` if the results are paged, and you must either return a cursor to get the next set of items or `undefined` if there are no more items.

**Comment by alekseyt** [comment]
>And if you have ideas on other types of information you'd like to see in this view, let us know!

This is really good stuff. This is probably not the point of this feature but it would be great to see this developing over time into full commit history (like gitk) and/or history navigation (like switching back and forth between revisions of a file (+ blame (+ comment))).

Even without those it's a great stuff, one less reason to switch to terminal, great job.

**Comment by JohnstonCode** [comment]
When will this API be moved from proposed?

**Comment by JacksonKearl** [comment]
Mentioned over here https://github.com/microsoft/vscode/issues/94155#issuecomment-607636707 but putting it in the main `api-proposal` too for visibility:

I think perhaps this would be easier to implement if it were like: 


```ts
interface TimelineProvider {
   /* existing properties */
   
  // Get most recent `items` timelime items, starting after `start` if defined
  provideTimeline(uri: Uri, options: TimelineOptions, token: CancellationToken, items: number, start?: string): ProviderResult<Timeline>

  // Refresh items up to and including `cursor`
  updateTimeline?(uri: Uri, options: TimelineOptions, token: CancellationToken, cursor: string): ProviderResult<Timeline>

  // Not sure what this does tbh. Maybe not needed at all?
  fetchEntireTimeline?(...)
}
```

This way extensions can implement a minimal subset of the functionality and we can detect that in core and work around it. For instance, providers can minimally implement just `provideTimeline`. In that case, if they trigger a refresh the user will just get a single page back. If they go on to implement `updateTimeline`, then the user can get their full set of results back. I'm not really sure what the use case is for `fetchEntireTimeline`, but similar for that.

Right now when implementing one needs to perform a lot of runtime type checking and it's not clear to the implementor what the different cursor overloads are. If the method names made clear what the results they should return, rather than one needing to know how to associate cursor types with runtime behaviour, it'd be a lot more clear how to implement.

**Comment by eamodio** [comment]
@JohnstonCode It might be this iteration, but more likely the next one.

@JacksonKearl I'm still really not in favor of something like this -- to me it just actually increases the complexity and when fully implementing it I would just have all those methods delegate down to the same underlying method.

The way it is now, a Timeline provider is pretty much in control over what they implement. If they don't want to implement paging or cursor, they can just ignore them and never return a `paging` object in the result. And the `limit` properties are guidelines, not absolute, so even if we ask for 20 items, and you always return all of your items, that is fine -- we will handle it appropriately.

I am thinking about a way to ease some of the confusion with `cursor` and `limit`. Because at a basic level, the `cursor` defines a starting point, and `limit` defines and ending point. So hopefully with some minor terminology changes, or slight shape tweaks this will become much clearer.

**Comment by eamodio** [comment]
📢 Anyone currently using/trying this API or planning on it please join us in the April Extension Authors call: https://github.com/microsoft/vscode/issues/96927. We will be talking about the Timeline API and would love your feedback as we plan to finalize the API.

**Comment by mjcrouch** [comment]
Hi - I'm implementing this for perforce - is there a way to identify that a user has explicitly requested a refresh?

I don't really want to hit the server for the file log every time we switch tabs (hence I've implemented a cache), but I would think if someone has pressed the refresh button then we would want to force an update to the state.

For example, if another person has submitted the file the user is working on, this is totally asynchronous in perforce and we don't really have a way of detecting that this has happened, as a user I wouldn't necessarily expect to immediately see that revision, but I would want to be able to click on the refresh button and see that new revision appear.

**Comment by joaomoreno** [comment]
@eamodio Since you're out on vacation, I'll move this forward.

**Comment by joaomoreno** [comment]
@eamodio Since you're out on vacation, I'll move this forward again. 😆

**Comment by InTheCloudDan** [comment]
This is just feedback from my very initial usage of the API, if `detail` supported `MarkdownString` similar to how the Treeview does now, for our use case it would allow us to show much more feature rich contextual info.

**Comment by eamodio** [comment]
@InTheCloudDan Can you please open a new issue for that -- I think it is a great idea.

**Comment by gjsjohnmurray** [comment]
@lszomoru can this be considered for finalization?

**Comment by JacksonKearl** [comment]
@gjsjohnmurray the remaining unknown is in dealing with workspace scoped timeline events. The timeline view would ideally have a way to present a full git log of the open repo(s), but right now it only works for single resources. My understanding is that we don't want to move forward with finalization until we have an answer for that.

**Comment by kraabrsg** [comment]
Hi, are there any plans to bring this into a release Version, we would eagerly like to use it.

Thanks!

**Comment by gjsjohnmurray** [comment]
Ahead of finalization of this API maybe revisit #147304 which acknowledges it's currently exposing two undocumented `TimelineOptions` properties `cacheResult` and `resetCache` to extensions.

**Comment by InTheCloudDan** [comment]
Any update, a timeline API would be helpful.

**Comment by gjsjohnmurray** [comment]
> [@gjsjohnmurray](https://github.com/gjsjohnmurray) the remaining unknown is in dealing with workspace scoped timeline events. The timeline view would ideally have a way to present a full git log of the open repo(s), but right now it only works for single resources. My understanding is that we don't want to move forward with finalization until we have an answer for that.

In the light of https://github.com/microsoft/vscode/issues/213935#issuecomment-2527628519 does "Source Control Graph" now meet this need? If so, please finalize the timeline API.

**Comment by gjsjohnmurray** [comment]
> In the light of [#213935 (comment)](https://github.com/microsoft/vscode/issues/213935#issuecomment-2527628519) does "Source Control Graph" now meet this need? If so, please finalize the timeline API.

@lramos15 is there any hope of this API getting finalized? As an extension developer I find it demotivating when a proposed API such as this seems to make no progress towards finalization. In this case it's existed for more than 5 years now but we still can't use it in extensions we want to publish on Marketplace.

**Comment by gjsjohnmurray** [comment]
@lramos15 please provide an update on this.

**Comment by kraabrsg** [comment]
@lramos15 please give an update on this. The focus seems to be on copilot only. Please focus on features / fixes / performance. Thanks.

**Comment by breathe67** [comment]
My upvote on this too.

**Comment by VKrishna04** [comment]
Please add this so that we can improve and optimise extensions like [Global Save State](https://marketplace.visualstudio.com/items?itemName=VKrishna04.global-save-state)

**Comment by iwangbowen** [comment]
Is it available? My extension needs this

**Comment by gjsjohnmurray** [comment]
Extension developers waiting for this API to be finalized may like to comment or upvote at https://github.com/microsoft/vscode-discussions/discussions/3033.

**Comment by gjsjohnmurray** [comment]
> [Logan Ramos (@lramos15)](https://github.com/lramos15) what can I do to help this make progress?

Reiterating my offer to help from 12 months ago.
